File: output/GPT4o-Mini/xCodeEval_4000/Java/Python/258f7a3ddf1d43ec0b7e14c602788767.py
import sys
from collections import defaultdict
from sortedcontainers import SortedSet

class F:
    def __init__(self):
        self.n = 0
        self.m = 0
        self.arr = []
        self.required = 0
        self.type_days = defaultdict(SortedSet)

    def solver(self):
        self.n, self.m = map(int, sys.stdin.readline().split())
        self.arr = list(map(int, sys.stdin.readline().split()))
        self.required = sum(self.arr)

        for _ in range(self.m):
            d, typ = map(int, sys.stdin.readline().split())
            if self.arr[typ - 1] <= 0:
                continue
            self.type_days[typ].add(d)

        lo = self.required
        hi = 2 * self.required
        valid = hi

        while lo <= hi:
            mid = (lo + hi) // 2
            if self.is_enough(mid):
                valid = mid
                hi = mid - 1
            else:
                lo = mid + 1

        print(valid)

    def is_enough(self, day):
        total_coin = day
        req = self.required

        ddays = [0] * (day + 1)
        for typ in self.type_days.keys():
            last_day = self.type_days[typ].bisect_right(day) - 1
            if last_day >= 0:
                ddays[self.type_days[typ][last_day]] += self.arr[typ - 1]

        removed = 0
        for i in range(1, day + 1):
            coins = i - removed
            to_remove = min(coins, ddays[i])
            removed += to_remove
            total_coin -= to_remove
            req -= to_remove

        return req * 2 <= total_coin

if __name__ == "__main__":
    F().solver()

File: output/GPT4o-Mini/xCodeEval_4000/Java/Go/258f7a3ddf1d43ec0b7e14c602788767.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type F struct {
	n, m      int
	arr       []int
	required  int
	typeDays  map[int][]int
}

func (f *F) solver() {
	f.required = sum(f.arr)
	f.typeDays = make(map[int][]int)

	for i := 0; i < f.m; i++ {
		var d, typ int
		fmt.Scan(&d, &typ)
		if f.arr[typ-1] <= 0 {
			continue
		}
		f.typeDays[typ] = append(f.typeDays[typ], d)
	}

	lo, hi := f.required, 2*f.required
	valid := hi
	for lo <= hi {
		mid := (lo + hi) / 2
		if f.isEnough(mid) {
			valid = mid
			hi = mid - 1
		} else {
			lo = mid + 1
		}
	}
	fmt.Println(valid)
}

func (f *F) isEnough(day int) bool {
	totalCoin := day
	req := f.required

	ddays := make([]int, day+1)
	for typ, days := range f.typeDays {
		sort.Ints(days)
		for _, d := range days {
			if d <= day {
				ddays[d] += f.arr[typ-1]
			}
		}
	}

	removed := 0
	for i := 1; i <= day; i++ {
		coins := i - removed
		toRemove := min(coins, ddays[i])
		removed += toRemove
		totalCoin -= toRemove
		req -= toRemove
	}

	return req*2 <= totalCoin
}

func sum(arr []int) int {
	total := 0
	for _, v := range arr {
		total += v
	}
	return total
}

func min(ins ...int) int {
	minVal := ins[0]
	for _, v := range ins[1:] {
		if v < minVal {
			minVal = v
		}
	}
	return minVal
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	var f F
	fmt.Fscan(reader, &f.n, &f.m)
	f.arr = make([]int, f.n)
	for i := 0; i < f.n; i++ {
		fmt.Fscan(reader, &f.arr[i])
	}
	f.solver()
}
